### ISSUE TITLE
Reader example should be filled

### ISSUE THREAD

**Issue by CJ-Wright** [NSLS-II/bluesky#796]
The `Reader` example should give its results as filled.

## Expected Behavior
Since `Reader` doesn't fill if you make an image in a reader when passed to `LiveImage` it errors as their neither a `filled` key nor a `Broker` instance which harbors the data.

**Comment by danielballan** [comment]
Sorry, I don't quite follow this. Can you share a minimal example?

**Comment by CJ-Wright** [comment]
Run the live image callback against a reader which generates 10x10 arrays. The callback will fail as the result is filled but the filled dict is not set. Also it doesn't have a databroker to get the info from (not that really matters in this example

**Comment by danielballan** [comment]
Can you share a minimal example with code?

**Comment by CJ-Wright** [comment]
Sorry, I wrote that on the train on my phone.

```python
In [4]: from bluesky.examples import Reader
   ...: from bluesky.callbacks.broker import LiveImage
   ...: import numpy as np
   ...: from bluesky.run_engine import RunEngine
   ...: from bluesky.plans import count
   ...: 
   ...: r = Reader('reader', fields={'reader': lambda: np.random.random((10, 10))})
   ...: RE = RunEngine({})
   ...: RE.subscribe(LiveImage('reader'))
   ...: RE(count([r], num=5))
```
gave
```python
/home/christopher/mc/envs/dp_dev/lib/python3.5/site-packages/matplotlib/axes/_base.py:2917: UserWarning: Attempting to set identical left==right results
in singular transformations; automatically expanding.
left=-0.5, right=-0.5
  'left=%s, right=%s') % (left, right))
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-4-85cb5411dd88> in <module>()
      8 RE = RunEngine({})
      9 RE.subscribe(LiveImage('reader'))
---> 10 RE(count([r], num=5))

/home/christopher/dev/bluesky/bluesky/run_engine.py in __call__(self, plan, subs, raise_if_interrupted, **metadata_kw)
    644                     # it (unless it is a canceled error)
    645                     if exc is not None:
--> 646                         raise exc
    647 
    648             if raise_if_interrupted and self._interrupted:

/home/christopher/mc/envs/dp_dev/lib/python3.5/asyncio/tasks.py in _step(***failed resolving arguments***)
    237                 # We use the `send` method directly, because coroutines
    238                 # don't have `__iter__` and `__next__` methods.
--> 239                 result = coro.send(None)
    240             else:
    241                 result = coro.throw(exc)

/home/christopher/dev/bluesky/bluesky/run_engine.py in _run(self)
   1084             self.log.error("Run aborted")
   1085             self.log.error("%r", err)
-> 1086             raise err
   1087         finally:
   1088             # Some done_callbacks may still be alive in other threads.

/home/christopher/dev/bluesky/bluesky/run_engine.py in _run(self)
    982                         resp = self._response_stack.pop()
    983                         try:
--> 984                             msg = self._plan_stack[-1].send(resp)
    985                         # We have exhausted the top generator
    986                         except StopIteration:

/home/christopher/dev/bluesky/bluesky/plans.py in count(detectors, num, delay, md)
   2081         return (yield from infinite_plan())
   2082     else:
-> 2083         return (yield from finite_plan())
   2084 
   2085 

/home/christopher/dev/bluesky/bluesky/utils.py in dec_inner(*inner_args, **inner_kwargs)
    920                 plan = gen_func(*inner_args, **inner_kwargs)
    921                 plan = wrapper(plan, *args, **kwargs)
--> 922                 return (yield from plan)
    923             return dec_inner
    924         return dec

/home/christopher/dev/bluesky/bluesky/plans.py in stage_wrapper(plan, devices)
   1609         return (yield from plan)
   1610 
-> 1611     return (yield from finalize_wrapper(inner(), unstage_devices()))
   1612 
   1613 

/home/christopher/dev/bluesky/bluesky/plans.py in finalize_wrapper(plan, final_plan, pause_for_debug)
   1110     cleanup = True
   1111     try:
-> 1112         ret = yield from plan
   1113     except GeneratorExit:
   1114         cleanup = False

/home/christopher/dev/bluesky/bluesky/plans.py in inner()
   1607     def inner():
   1608         yield from stage_devices()
-> 1609         return (yield from plan)
   1610 
   1611     return (yield from finalize_wrapper(inner(), unstage_devices()))

/home/christopher/dev/bluesky/bluesky/utils.py in dec_inner(*inner_args, **inner_kwargs)
    920                 plan = gen_func(*inner_args, **inner_kwargs)
    921                 plan = wrapper(plan, *args, **kwargs)
--> 922                 return (yield from plan)
    923             return dec_inner
    924         return dec

/home/christopher/dev/bluesky/bluesky/plans.py in run_wrapper(plan, md)
    912     rs_uid = yield from open_run(md)
    913     yield from finalize_wrapper(plan,
--> 914                                 close_run())
    915     return rs_uid
    916 

/home/christopher/dev/bluesky/bluesky/plans.py in finalize_wrapper(plan, final_plan, pause_for_debug)
   1110     cleanup = True
   1111     try:
-> 1112         ret = yield from plan
   1113     except GeneratorExit:
   1114         cleanup = False

/home/christopher/dev/bluesky/bluesky/plans.py in finite_plan()
   2050             now = time.time() # Intercept the flow in its earliest moment.
   2051             yield Msg('checkpoint')
-> 2052             yield from trigger_and_read(detectors)
   2053             try:
   2054                 d = next(delay)

/home/christopher/dev/bluesky/bluesky/plans.py in trigger_and_read(devices, name)
   1887 
   1888     return (yield from rewindable_wrapper(inner_trigger_and_read(),
-> 1889                                           rewindable))
   1890 
   1891 

/home/christopher/dev/bluesky/bluesky/plans.py in rewindable_wrapper(plan, rewindable)
   1321                                             restore_rewindable()))
   1322     else:
-> 1323         return (yield from plan)
   1324 
   1325 

/home/christopher/dev/bluesky/bluesky/plans.py in inner_trigger_and_read()
   1882             if reading is not None:
   1883                 ret.update(reading)
-> 1884         yield from save()
   1885         return ret
   1886 

/home/christopher/dev/bluesky/bluesky/plans.py in save()
    299     :func:`bluesky.plans.create`
    300     """
--> 301     return (yield Msg('save'))
    302 
    303 

/home/christopher/dev/bluesky/bluesky/run_engine.py in _run(self)
   1033                         # exceptions (coming in via throw) can be
   1034                         # raised
-> 1035                         response = yield from coro(msg)
   1036                     # special case `CancelledError` and let the outer
   1037                     # exception block deal with it.

/home/christopher/dev/bluesky/bluesky/run_engine.py in _save(self, msg)
   1536                    time=ttime.time(), data=data, timestamps=timestamps,
   1537                    seq_num=seq_num, uid=event_uid, filled=filled)
-> 1538         yield from self.emit(DocumentNames.event, doc)
   1539         self.log.debug("Emitted Event with data keys %r (uid=%r)", data.keys(),
   1540                        event_uid)

/home/christopher/mc/envs/dp_dev/lib/python3.5/asyncio/coroutines.py in coro(*args, **kw)
    208         @functools.wraps(func)
    209         def coro(*args, **kw):
--> 210             res = func(*args, **kw)
    211             if (futures.isfuture(res) or inspect.isgenerator(res) or
    212                 isinstance(res, CoroWrapper)):

/home/christopher/dev/bluesky/bluesky/run_engine.py in emit(self, name, doc)
   2069         "Process blocking callbacks and schedule non-blocking callbacks."
   2070         jsonschema.validate(doc, schemas[name])
-> 2071         self.dispatcher.process(name, doc)
   2072 
   2073 

/home/christopher/dev/bluesky/bluesky/run_engine.py in process(self, name, doc)
   2089         doc : dict
   2090         """
-> 2091         exceptions = self.cb_registry.process(name, name.name, doc)
   2092         for exc, traceback in exceptions:
   2093             warn("A %r was raised during the processing of a %s "

/home/christopher/dev/bluesky/bluesky/utils.py in process(self, sig, *args, **kwargs)
    276             for cid, func in list(self.callbacks[sig].items()):
    277                 try:
--> 278                     func(*args, **kwargs)
    279                 except ReferenceError:
    280                     self._remove_proxy(func)

/home/christopher/dev/bluesky/bluesky/utils.py in __call__(self, *args, **kwargs)
    357             mtd = self.func
    358         # invoke the callable and return the result
--> 359         return mtd(*args, **kwargs)
    360 
    361     def __eq__(self, other):

/home/christopher/dev/bluesky/bluesky/callbacks/core.py in __call__(self, name, doc)
     19     def __call__(self, name, doc):
     20         "Dispatch to methods expecting particular doc types."
---> 21         return getattr(self, name)(doc)
     22 
     23     def event(self, doc):

/home/christopher/dev/bluesky/bluesky/callbacks/broker.py in event(self, doc)
     89 
     90     def event(self, doc):
---> 91         super().event(doc)
     92         data = doc['data'][self.field]
     93         self.update(data)

/home/christopher/dev/bluesky/bluesky/callbacks/broker.py in event(self, doc)
     41         if fields:
     42             if self.db is None:
---> 43                 raise RuntimeError('Either the data must be pre-loaded or '
     44                                    'a Broker instance must be provided '
     45                                    'via the db parameter of '

RuntimeError: Either the data must be pre-loaded or a Broker instance must be provided via the db parameter of BrokerCallbackBase.
```

**Comment by danielballan** [comment]
Thanks, that's clearer. I think the change should go into BrokerCallbackBase. It should first check whether data is external in the first place before going in to check whether it was filled.

**Comment by CJ-Wright** [comment]
Hmm, that is one route. I can try to put in a PR to make the needed change to check if the data is external.

**Comment by danielballan** [comment]
Great, thanks.

**Comment by danielballan** [comment]
Closed by #803
